core/src/main/java/com/linecorp/armeria/internal/common/stream/HttpDataDeframer.java

```diff
@@ -0,0 +1,417 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayDeque;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Processor;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.stream.DefaultStreamMessage;
+import com.linecorp.armeria.common.util.Exceptions;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * A skeletal {@link Processor} implementation that decodes a stream of {@link HttpData}s to N objects.
+ */
+public abstract class HttpDataDeframer<T> extends DefaultStreamMessage<T> implements Processor<HttpData, T> {
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<HttpDataDeframer, Subscription> upstreamUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(HttpDataDeframer.class, Subscription.class, "upstream");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<HttpDataDeframer> initializedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(HttpDataDeframer.class, "initialized");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<HttpDataDeframer> subscribedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(HttpDataDeframer.class, "subscribed");
+
+    private final Queue<ByteBuf> queue;
+    private final HttpDataReader reader;
+    private final EventExecutor eventLoop;
+    private final ByteBufAllocator alloc;
+
+    @Nullable
+    private volatile Subscription upstream;
+    private volatile boolean isCancelled;
+    private volatile int initialized;
+    private volatile int subscribed;
+
+    /**
+     * Returns a new {@link HttpDataDeframer} with the specified {@link EventExecutor} and
+     * {@link ByteBufAllocator}.
+     */
+    protected HttpDataDeframer(EventExecutor eventLoop, ByteBufAllocator alloc) {
+        requireNonNull(eventLoop, "eventLoop");
+        requireNonNull(alloc, "alloc");
+        queue = new ArrayDeque<>();
+        reader = new HttpDataReader();
+        this.eventLoop = eventLoop;
+        this.alloc = alloc;
+    }
+
+    /**
+     * Decodes a stream of {@link HttpData}s to N objects.
+     * This method will be called whenever an {@link HttpData} is signaled from {@link Publisher}.
+     *
+     * <p>Note that The {@link HttpData}s which was fully consumed will be cleaned up automatically
+     * so that the {@link Queue} does not grow up infinitely.
```

Comment:
> so that the Queue does not grow up infinitely.

Maybe too much information to users because users don't care about the underlying data structure?